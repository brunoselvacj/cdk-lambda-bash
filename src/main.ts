import * as fs from 'fs';
import * as path from 'path';
import {
  CfnOutput, CustomResource, Duration, AssetStaging,
  aws_iam as iam,
  aws_lambda as lambda,
  aws_logs as logs,
  custom_resources as cr,
} from 'aws-cdk-lib';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { Construct } from 'constructs';

export interface BashExecFunctionProps {
  /**
   * The path of the shell script to be executed.
   */
  readonly script: string;

  /**
   * The path of your custom dockerfile.
   */
  readonly dockerfile?: string;

  /**
   * Lambda environment variables.
   */
  readonly environment?: { [key: string]: string };

  /**
   * The function execution time (in seconds) after which Lambda terminates the function.
   * Because the execution time affects cost, set this value based on the function's expected execution time.
   * @default - Duration.seconds(60)
   * */
  readonly timeout?: Duration;

  /**
   * Custom lambda execution role.
   *
   * @default - auto generated role.
   */
  readonly role?: iam.IRole;

  /**
   * Custom lambda execution vpc.
   *
   * @default - auto generated vpc.
   */
  readonly vpc?: ec2.IVpc;

  /**
   * Custom lambda execution vpcSubnets.
   *
   * @default - auto generated vpcSubnets.
   */
  readonly vpcSubnets?: ec2.SubnetSelection;

  /**
   * Custom lambda execution securityGroups.
   *
   * @default - auto generated securityGroups.
   */
  readonly securityGroups?: ec2.ISecurityGroup[];
}

export interface RunOps {
  /**
   * whether to run the lambda function again on the provider update
   *
   * @default false;
   */
  readonly runOnUpdate?: boolean;
}

export class BashExecFunction extends Construct {
  readonly handler: lambda.DockerImageFunction;
  constructor(scope: Construct, id: string, props: BashExecFunctionProps) {
    super(scope, id);

    if (props?.dockerfile && fs.existsSync(props?.dockerfile)) {
      // Copy your Dockerfile to Dockerfile.custom.
      fs.copyFileSync(props?.dockerfile, path.join(__dirname, '../docker.d/Dockerfile.custom'));
    }
    const dockerDirPath = path.join(__dirname, '../docker.d');
    const scriptPath = props.script;

    // copy the user script to the docker.d directory as main.sh so we can bundle it up into a new docker image
    const mainFile = path.join(dockerDirPath, '/main.sh');
    fs.copyFileSync(scriptPath, mainFile);

    this.handler = new lambda.DockerImageFunction(this, 'BashExecFunction', {
      code: lambda.DockerImageCode.fromImageAsset(dockerDirPath, {
        file: props?.dockerfile && fs.existsSync(props?.dockerfile) ? 'Dockerfile.custom': undefined,
      }),
      timeout: props.timeout ?? Duration.seconds(60),
      logRetention: logs.RetentionDays.ONE_DAY,
      environment: props.environment,
      role: props.role,
      vpc: props.vpc,
      vpcSubnets: props.vpcSubnets,
      securityGroups: props.securityGroups,
    });
    new CfnOutput(this, 'LogGroup', { value: this.handler.logGroup.logGroupName });
  }
  public run(ops: RunOps = {}): CustomResource {
    const onEvent = new lambda.DockerImageFunction(this, 'OnEventHandler', {
      code: lambda.DockerImageCode.fromImageAsset(path.join(__dirname, '../docker.d'), {
        cmd: ['function.sh.onEvent'],
      }),
      environment: {
        LAMBDA_FUNCTION_ARN: this.handler.functionArn,
      },
      timeout: Duration.seconds(60),
    });
    const myProvider = new cr.Provider(this, 'MyProvider', {
      onEventHandler: onEvent,
    });

    const staging = new AssetStaging(this, 'Staging', {
      sourcePath: path.join(__dirname, '../docker.d'),
    });

    const resource = new CustomResource(this, 'RunLambdaBash', {
      resourceType: 'Custom::RunLambdaBash',
      serviceToken: myProvider.serviceToken,
      properties: {
        assetHash: ops.runOnUpdate ? staging.assetHash : undefined,
      },
    });
    this.handler.grantInvoke(onEvent.grantPrincipal);

    return resource;
  }

}
