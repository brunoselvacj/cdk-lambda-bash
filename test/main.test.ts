import * as path from 'path';
import {
  App, Stack,
  aws_iam as iam,
  aws_s3 as s3,
} from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { BashExecFunction } from '../src';
import { IntegTesting } from '../src/integ.default';

test('integ snapshot validation', () => {
  const integ = new IntegTesting();
  integ.stack.forEach(stack => {
    const t = Template.fromStack(stack);
    expect(t).toMatchSnapshot();
  });
});


test('re-execution on assets update', () => {
  const app = new App();
  const stack = new Stack(app, 'test-stack');
  new BashExecFunction(stack, 'Demo', {
    script: path.join(__dirname, '../demo.sh'),
  }).run({ runOnUpdate: true });
  const t = Template.fromStack(stack);
  t.hasResourceProperties('Custom::RunLambdaBash', {
    assetHash: '9ddb8373fd34c96b5b86bd0d72e8fc9e3642733fbb4bf6d84f5c12461a9c71aa',
  });
});

test('run should return custom resource', () => {
  const app = new App();
  const stack = new Stack(app, 'test-stack');
  const fn = new BashExecFunction(stack, 'Demo', {
    script: path.join(__dirname, '../demo.sh'),
  });
  const cr = fn.run({ runOnUpdate: true });
  const bucket = new s3.Bucket(stack, 'Bucket');
  const t = Template.fromStack(stack);
  cr.node.addDependency(bucket);
  t.resourceCountIs('Custom::RunLambdaBash', 1);
  // to-fix: to check the dependency with cdk-assertions
  // t.hasResource('Custom::RunLambdaBash', {
  //       DependsOn: ['Bucket83908E77'],
  // })
});

test('custom lambda execution role', () => {
  const app = new App();
  const stack = new Stack(app, 'test-stack');
  const customRole = new iam.Role(stack, 'Role', {
    roleName: 'customRole',
    assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
  });
  new BashExecFunction(stack, 'Demo', {
    script: path.join(__dirname, '../demo.sh'),
    role: customRole,
  });
  const t = Template.fromStack(stack);
  t.hasResourceProperties('AWS::Lambda::Function', {
    Role: {
      'Fn::GetAtt': [
        'Role1ABCC5F0',
        'Arn',
      ],
    },
  });
});
