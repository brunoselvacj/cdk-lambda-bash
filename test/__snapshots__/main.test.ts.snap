// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integ snapshot validation 1`] = `
Object {
  "Outputs": Object {
    "DemoLogGroupADB0CC2E": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "DemoBashExecFunctionLogRetention092AB8A2",
          "LogGroupName",
        ],
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "DemoBashExecFunction72E60776": Object {
      "DependsOn": Array [
        "DemoBashExecFunctionServiceRole911B396A",
      ],
      "Properties": Object {
        "Code": Object {
          "ImageUri": Object {
            "Fn::Sub": "\${AWS::AccountId}.dkr.ecr.\${AWS::Region}.\${AWS::URLSuffix}/cdk-hnb659fds-container-assets-\${AWS::AccountId}-\${AWS::Region}:c0afe9769c244fad44491937969e647fe3596cbcbfbf4ce26da9136be2868312",
          },
        },
        "Environment": Object {
          "Variables": Object {
            "FOO": "BAR",
          },
        },
        "PackageType": "Image",
        "Role": Object {
          "Fn::GetAtt": Array [
            "DemoBashExecFunctionServiceRole911B396A",
            "Arn",
          ],
        },
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "DemoBashExecFunctionLogRetention092AB8A2": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "DemoBashExecFunction72E60776",
              },
            ],
          ],
        },
        "RetentionInDays": 1,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "DemoBashExecFunctionServiceRole911B396A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DemoMyProviderframeworkonEventB6648094": Object {
      "DependsOn": Array [
        "DemoMyProviderframeworkonEventServiceRoleDefaultPolicy5AB8994E",
        "DemoMyProviderframeworkonEventServiceRole5F456F12",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "c691172cdeefa2c91b5a2907f9d81118e47597634943344795f1a844192dd49c.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (lambda-bash-dev/Demo/MyProvider)",
        "Environment": Object {
          "Variables": Object {
            "USER_ON_EVENT_FUNCTION_ARN": Object {
              "Fn::GetAtt": Array [
                "DemoOnEventHandler90C6007F",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": Object {
          "Fn::GetAtt": Array [
            "DemoMyProviderframeworkonEventServiceRole5F456F12",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "DemoMyProviderframeworkonEventServiceRole5F456F12": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DemoMyProviderframeworkonEventServiceRoleDefaultPolicy5AB8994E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "DemoOnEventHandler90C6007F",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DemoMyProviderframeworkonEventServiceRoleDefaultPolicy5AB8994E",
        "Roles": Array [
          Object {
            "Ref": "DemoMyProviderframeworkonEventServiceRole5F456F12",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DemoOnEventHandler90C6007F": Object {
      "DependsOn": Array [
        "DemoOnEventHandlerServiceRoleDefaultPolicy8ABF7FCE",
        "DemoOnEventHandlerServiceRole2506682B",
      ],
      "Properties": Object {
        "Code": Object {
          "ImageUri": Object {
            "Fn::Sub": "\${AWS::AccountId}.dkr.ecr.\${AWS::Region}.\${AWS::URLSuffix}/cdk-hnb659fds-container-assets-\${AWS::AccountId}-\${AWS::Region}:c0afe9769c244fad44491937969e647fe3596cbcbfbf4ce26da9136be2868312",
          },
        },
        "Environment": Object {
          "Variables": Object {
            "LAMBDA_FUNCTION_ARN": Object {
              "Fn::GetAtt": Array [
                "DemoBashExecFunction72E60776",
                "Arn",
              ],
            },
          },
        },
        "ImageConfig": Object {
          "Command": Array [
            "function.sh.onEvent",
          ],
        },
        "PackageType": "Image",
        "Role": Object {
          "Fn::GetAtt": Array [
            "DemoOnEventHandlerServiceRole2506682B",
            "Arn",
          ],
        },
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "DemoOnEventHandlerServiceRole2506682B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DemoOnEventHandlerServiceRoleDefaultPolicy8ABF7FCE": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "DemoBashExecFunction72E60776",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DemoOnEventHandlerServiceRoleDefaultPolicy8ABF7FCE",
        "Roles": Array [
          Object {
            "Ref": "DemoOnEventHandlerServiceRole2506682B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DemoRunLambdaBash1B9713A5": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "DemoMyProviderframeworkonEventB6648094",
            "Arn",
          ],
        },
        "assetHash": "9ddb8373fd34c96b5b86bd0d72e8fc9e3642733fbb4bf6d84f5c12461a9c71aa",
      },
      "Type": "Custom::RunLambdaBash",
      "UpdateReplacePolicy": "Delete",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "b120b13d9d868c7622e7db1b68bae4c0f82ffd0227b8c15f2cef38e186ff3827.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
